internal/consensus/state.go

```diff
@@ -2412,3 +2412,31 @@ func repairWalFile(src, dst string) error {
 
 	return nil
 }
+
+// proposerWaitTime determines how long the proposer should wait to propose its next block.
+// Block times must be monotonically increasing, so if the block time of the previous
+// block is larger than the proposer's current time, then the proposer will sleep
+// until its local clock exceeds the previous block time.
+func proposerWaitTime(lt tmtime.Source, h types.Header) time.Duration {
+	t := lt.Now()
+	if t.After(h.Time) {
+		return 0
+	}
+	return h.Time.Sub(t)
+}
+
+// proposalStepWaitingTime determines how long a validator should wait for a block
+// to be sent from a proposer. This duration is determined as a result of the
+// previous block's time as well as by the Accuracy and MsgDelay timestamp parameters.
+// The validator's and the proposer's clocks should differ from eachother by at most
+// 2 * Accuracy and the proposal should take at most MsgDelay to arrive at the validator.
+// The validator must therefore wait at least 2 * Accuracy + MsgDelay for the proposal
+// to arrive.
```

Comment:
```suggestion
// proposalStepWaitingTime determines how long a validator should wait for a block
// to be sent from a proposer. If zero, the validator should not wait.
//
// This duration is determined as a result of the previous block's time as well as
// by the Accuracy and MsgDelay timestamp parameters.
// The validator's and the proposer's clocks should differ from each other by at most
// 2 * Accuracy and the proposal should take at most MsgDelay to arrive at the validator.
// The validator must therefore wait at least 2 * Accuracy + MsgDelay for the proposal
// to arrive.
```

Can we say something about what happens if the "should" is violated? (I.e., in the event the clocks differ by more than that, or the actual delay is longer)

Reply:
Yeah by:
> If zero, the validator should not wait.
You mean that the validator should prevote nil and advance to the next step

Reply:
I've reworded to hopefully more accurately reflect that goal of this function.